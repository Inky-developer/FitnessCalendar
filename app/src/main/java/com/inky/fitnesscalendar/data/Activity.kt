package com.inky.fitnesscalendar.data

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import java.util.Date

@Entity
data class Activity(
    @PrimaryKey val uid: Int? = null,
    @ColumnInfo(name = "type") val type: ActivityType,
    @ColumnInfo(name = "vehicle") val vehicle: Vehicle? = null,
    @ColumnInfo(name = "description") val description: String = "",
    @ColumnInfo(name = "start_time") val startTime: Date,
    @ColumnInfo(name = "end_time") val endTime: Date,
) {
    fun clean() = copy(
        vehicle = if (type.hasVehicle) vehicle else null,
        endTime = if (type.hasDuration) endTime else startTime
    )
}
