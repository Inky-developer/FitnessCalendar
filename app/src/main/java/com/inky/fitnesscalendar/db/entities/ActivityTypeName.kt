package com.inky.fitnesscalendar.db.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey

/**
 * Maps a string to an activity type.
 * For example, when importing gpx files, this will be used to convert the activity type
 * specified in the gpx file to an [ActivityType]
 */
@Entity(
    foreignKeys = [
        ForeignKey(
            entity = ActivityType::class,
            parentColumns = arrayOf("uid"),
            childColumns = arrayOf("type_id"),
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [Index("type_id")]
)
data class ActivityTypeName(
    @PrimaryKey val name: String,
    @ColumnInfo(name = "type_id") val typeId: Int
)