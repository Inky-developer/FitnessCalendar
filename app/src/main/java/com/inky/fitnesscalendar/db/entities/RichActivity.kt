package com.inky.fitnesscalendar.db.entities

import androidx.room.Embedded
import androidx.room.Relation

data class RichActivity(
    @Embedded val activity: Activity,
    @Relation(parentColumn = "type_id", entityColumn = "uid")
    val type: ActivityType
) {
    init {
        assert(activity.typeId == type.uid) { "Inconsistent TypeActivity: type is $type, but id is ${activity.typeId}" }
    }

    fun clean() = copy(type = type, activity = activity.clean(type))
}