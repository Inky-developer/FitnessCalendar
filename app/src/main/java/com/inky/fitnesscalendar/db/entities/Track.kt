package com.inky.fitnesscalendar.db.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import com.inky.fitnesscalendar.data.gpx.GpxTrackPoint
import com.inky.fitnesscalendar.data.gpx.GpxTrackStats
import com.inky.fitnesscalendar.data.measure.Distance
import com.inky.fitnesscalendar.data.measure.Duration
import com.inky.fitnesscalendar.data.measure.Duration.Companion.until
import com.inky.fitnesscalendar.data.measure.HeartFrequency
import com.inky.fitnesscalendar.data.measure.Speed
import com.inky.fitnesscalendar.data.measure.Temperature
import kotlin.math.roundToLong

@Entity(
    foreignKeys = [
        ForeignKey(
            entity = Activity::class,
            parentColumns = ["uid"],
            childColumns = ["activity_id"],
            onDelete = ForeignKey.CASCADE,
        )
    ],
    indices = [Index("activity_id")]
)
data class Track(
    @PrimaryKey val uid: Int? = null,
    @ColumnInfo(name = "activity_id") val activityId: Int,
    @ColumnInfo(name = "points") val points: List<GpxTrackPoint>
) {
    val startTime get() = points.firstOrNull()?.time
    val endTime get() = points.lastOrNull()?.time

    private val totalDuration get() = endTime?.let { startTime?.until(it) }

    /**
     * Adds all relevant data of this track to the activity
     */
    fun addStatsToActivity(activity: Activity): Activity? {
        val stats = computeStatistics() ?: return null
        return activity.copy(
            startTime = startTime ?: return null,
            endTime = endTime ?: return null,
            distance = stats.totalDistance,
            movingDuration = stats.movingDuration,
            temperature = stats.averageTemperature,
            averageHeartRate = stats.averageHeartFrequency,
            maximalHeartRate = stats.maxHeartFrequency
        )
    }

    private fun computeStatistics(): GpxTrackStats? {
        val duration = totalDuration ?: return null

        val trackPointComputedData = computeTrackPointComputedData()

        val totalDistance =
            Distance(meters = trackPointComputedData.sumOf { it.distanceMeters }.roundToLong())
        val movingDuration = Duration(
            elapsedMs = trackPointComputedData
                .filter { it.speed.kmh > 3 }
                .sumOf { it.duration.elapsedMs }
        )

        val averageHeartRate = points.mapNotNull { it.heartFrequency?.bpm }.let {
            if (it.isEmpty()) {
                null
            } else {
                HeartFrequency(bpm = it.sum() / it.size)
            }
        }
        val maxHeartRate = points.mapNotNull { it.heartFrequency }.maxByOrNull { it }

        val averageTemperature = points
            .mapNotNull { it.temperature?.celsius }
            .let {
                if (it.isEmpty()) {
                    null
                } else {
                    Temperature(celsius = it.sum() / it.size)
                }
            }
        val minTemperature = points.mapNotNull { it.temperature }.minByOrNull { it }
        val maxTemperature = points.mapNotNull { it.temperature }.maxByOrNull { it }

        return GpxTrackStats(
            totalDistance = totalDistance,
            totalDuration = duration,
            movingDuration = movingDuration,
            averageHeartFrequency = averageHeartRate,
            maxHeartFrequency = maxHeartRate,
            averageTemperature = averageTemperature,
            minTemperature = minTemperature,
            maxTemperature = maxTemperature
        )
    }

    private fun computeTrackPointComputedData(): List<TrackPointComputedData> {
        val cachedDistanceResultsArray = FloatArray(3)
        val computedData = points
            .windowed(2)
            .map { (a, b) ->
                val distance = a.coordinate.distanceMeters(b.coordinate, cachedDistanceResultsArray)
                val duration = a.time.until(b.time)
                TrackPointComputedData(distance, duration)
            }

        return computedData
    }

    private data class TrackPointComputedData(val distanceMeters: Double, val duration: Duration) {
        val speed get() = Speed(metersPerSecond = distanceMeters / duration.elapsedSeconds)
    }
}
