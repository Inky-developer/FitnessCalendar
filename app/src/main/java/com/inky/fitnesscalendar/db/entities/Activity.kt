package com.inky.fitnesscalendar.db.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import com.inky.fitnesscalendar.data.EpochDay
import com.inky.fitnesscalendar.data.Feel
import com.inky.fitnesscalendar.data.ImageName
import com.inky.fitnesscalendar.data.Intensity
import com.inky.fitnesscalendar.data.Vehicle
import com.inky.fitnesscalendar.data.gpx.TrackSvg
import com.inky.fitnesscalendar.data.measure.Distance
import com.inky.fitnesscalendar.data.measure.Duration
import com.inky.fitnesscalendar.data.measure.Duration.Companion.until
import com.inky.fitnesscalendar.data.measure.HeartFrequency
import com.inky.fitnesscalendar.data.measure.Speed
import com.inky.fitnesscalendar.data.measure.Temperature
import com.inky.fitnesscalendar.data.measure.VerticalDistance
import com.inky.fitnesscalendar.data.measure.metersPerSecond
import com.inky.fitnesscalendar.util.toLocalDate
import kotlinx.serialization.json.Json
import java.util.Date

@Entity(
    foreignKeys = [
        ForeignKey(
            entity = ActivityType::class,
            parentColumns = arrayOf("uid"),
            childColumns = arrayOf("type_id"),
            onDelete = ForeignKey.RESTRICT
        ),
        ForeignKey(
            entity = Place::class,
            parentColumns = arrayOf("uid"),
            childColumns = arrayOf("place_id"),
            onDelete = ForeignKey.RESTRICT
        )
    ],
    indices = [Index("type_id"), Index("place_id")]
)
data class Activity(
    @PrimaryKey val uid: Int? = null,
    @ColumnInfo(name = "type_id") val typeId: Int,
    @ColumnInfo(name = "place_id") val placeId: Int? = null,
    @ColumnInfo(name = "vehicle") val vehicle: Vehicle? = null,
    @ColumnInfo(name = "description") val description: String = "",
    @ColumnInfo(name = "favorite", defaultValue = "false") val favorite: Boolean = false,

    @ColumnInfo(name = "image_name") val imageName: ImageName? = null,
    @ColumnInfo(name = "feel") val feel: Feel = Feel.Ok,
    @ColumnInfo(name = "intensity") val intensity: Intensity? = null,

    @ColumnInfo(name = "start_time", index = true) val startTime: Date,
    @ColumnInfo(name = "end_time") val endTime: Date = startTime,
    @ColumnInfo(name = "distance") val distance: Distance? = null,
    @ColumnInfo(name = "moving_duration") val movingDuration: Duration? = null,
    @ColumnInfo(name = "temperature") val temperature: Temperature? = null,
    @ColumnInfo(name = "average_heart_rate") val averageHeartRate: HeartFrequency? = null,
    @ColumnInfo(name = "maximal_heart_rate") val maximalHeartRate: HeartFrequency? = null,
    @ColumnInfo(name = "total_ascent") val totalAscent: VerticalDistance? = null,
    @ColumnInfo(name = "total_descent") val totalDescent: VerticalDistance? = null,

    @ColumnInfo(name = "track_preview") val trackPreview: SerializedTrackPreview? = null,
) {
    fun clean(type: ActivityType) = copy(
        placeId = if (type.hasPlace) placeId else null,
        vehicle = if (type.hasVehicle) vehicle else null,
        endTime = maxOf(if (type.hasDuration) endTime else startTime, startTime),
        distance = if (type.hasDistance) distance else null,
        intensity = if (type.hasIntensity) intensity else null,
    )

    val epochDay get() = EpochDay(startTime.toLocalDate().toEpochDay())

    val duration
        get() = startTime until endTime

    val averageSpeed: Speed?
        get() {
            if (duration.elapsedMs == 0L || distance == null) {
                return null
            }

            return (distance.meters / duration.elapsedSeconds).metersPerSecond()
        }

    val averageMovingSpeed: Speed?
        get() {
            if (movingDuration == null || distance == null) {
                return averageSpeed
            }

            return (distance.meters / movingDuration.elapsedSeconds).metersPerSecond()
        }

    @JvmInline
    value class SerializedTrackPreview(private val inner: String) {
        val value get() = inner

        fun toTrackSvg(): TrackSvg = Json.decodeFromString(value)

        companion object {
            fun assumeValid(value: String) = SerializedTrackPreview(value)

            fun TrackSvg.serialize() = SerializedTrackPreview(Json.encodeToString(this))
        }
    }
}
