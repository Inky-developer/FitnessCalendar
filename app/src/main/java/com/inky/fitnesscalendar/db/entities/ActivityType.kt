package com.inky.fitnesscalendar.db.entities

import android.content.Context
import android.os.Parcelable
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.inky.fitnesscalendar.data.ActivityCategory
import com.inky.fitnesscalendar.data.ActivityTypeColor
import com.inky.fitnesscalendar.data.Displayable
import kotlinx.parcelize.Parcelize

@Parcelize
@Entity
data class ActivityType(
    @PrimaryKey val uid: Int? = null,
    @ColumnInfo(name = "activity_category") val activityCategory: ActivityCategory,
    @ColumnInfo(name = "name") val name: String,
    @ColumnInfo(name = "emoji") val emoji: String,
    @ColumnInfo(name = "color") val color: ActivityTypeColor,
    @ColumnInfo(name = "has_vehicle") val hasVehicle: Boolean = false,
    @ColumnInfo(name = "has_duration") val hasDuration: Boolean = true,
    @ColumnInfo(name = "has_distance", defaultValue = "false") val hasDistance: Boolean = false,
    @ColumnInfo(name = "has_intensity", defaultValue = "false") val hasIntensity: Boolean = false,
) : Displayable, Parcelable {
    fun hasFeel() = hasDuration

    override fun getColor(context: Context) = context.getColor(color.colorId)

    override fun getText(context: Context) = name

    override fun getShortText() = emoji
}