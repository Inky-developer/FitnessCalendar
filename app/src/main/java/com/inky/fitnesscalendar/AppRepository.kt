package com.inky.fitnesscalendar

import android.content.Context
import android.util.Log
import androidx.compose.runtime.Immutable
import androidx.room.withTransaction
import com.inky.fitnesscalendar.data.EpochDay
import com.inky.fitnesscalendar.data.Vehicle
import com.inky.fitnesscalendar.data.activity_filter.ActivityFilter
import com.inky.fitnesscalendar.data.activity_filter.ActivityFilterChip
import com.inky.fitnesscalendar.data.activity_filter.ActivityFilterChip.Companion.toActivityFilterChip
import com.inky.fitnesscalendar.db.AppDatabase
import com.inky.fitnesscalendar.db.dao.ActivityDao
import com.inky.fitnesscalendar.db.dao.ActivityTypeDao
import com.inky.fitnesscalendar.db.dao.DayDao
import com.inky.fitnesscalendar.db.dao.FilterHistoryDao
import com.inky.fitnesscalendar.db.dao.PlaceDao
import com.inky.fitnesscalendar.db.dao.RecordingDao
import com.inky.fitnesscalendar.db.entities.Activity
import com.inky.fitnesscalendar.db.entities.ActivityType
import com.inky.fitnesscalendar.db.entities.Day
import com.inky.fitnesscalendar.db.entities.Place
import com.inky.fitnesscalendar.db.entities.Recording
import com.inky.fitnesscalendar.db.entities.RichActivity
import com.inky.fitnesscalendar.db.entities.RichRecording
import com.inky.fitnesscalendar.di.ActivityTypeOrder
import com.inky.fitnesscalendar.localization.LocalizationRepository
import com.inky.fitnesscalendar.util.hideRecordingNotification
import com.inky.fitnesscalendar.util.showRecordingNotification
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "AppRepository"

@Immutable
@Singleton
class AppRepository @Inject constructor(
    @ApplicationContext val context: Context,
    private val database: AppDatabase,
    private val activityDao: ActivityDao,
    private val recordingDao: RecordingDao,
    private val activityTypeDao: ActivityTypeDao,
    private val filterHistoryDao: FilterHistoryDao,
    private val dayDao: DayDao,
    private val placeDao: PlaceDao,
    val localizationRepository: LocalizationRepository
) {
    suspend fun loadAllActivities() = activityDao.loadActivities()

    fun getActivities(filter: ActivityFilter): Flow<List<RichActivity>> {
        // Use simpler query if no filters are needed
        if (filter.isEmpty()) {
            return activityDao.getActivities()
        }

        val searchVehicles = filter.text?.let { filterText ->
            Vehicle.entries.filter { vehicle ->
                context.getString(vehicle.nameId).contains(filterText, ignoreCase = true)
            }
        } ?: emptyList()

        val categories = filter.categories.map { it.toString() }

        val rangeDate = filter.range?.getDateRange()

        val hasDescription = filter.attributes.description.toBooleanOrNull()
        val hasFeel = filter.attributes.feel.toBooleanOrNull()
        val hasImage = filter.attributes.image.toBooleanOrNull()

        return activityDao.getFiltered(
            typeIds = filter.types.mapNotNull { it.uid },
            isTypesEmpty = filter.types.isEmpty(),
            categories = categories,
            isCategoriesEmpty = categories.isEmpty(),
            placeIds = filter.places.mapNotNull { it.uid },
            isPlacesEmpty = filter.places.isEmpty(),
            search = filter.text?.let { "%$it%" },
            searchVehicles = searchVehicles,
            start = rangeDate?.start,
            end = rangeDate?.end,
            hasDescription = hasDescription,
            hasFeel = hasFeel,
            hasImage = hasImage
        )
    }

    fun getMostRecentActivity() = activityDao.getMostRecentActivity()

    suspend fun saveActivity(activity: RichActivity) {
        activityDao.save(activity.clean().activity)
    }

    suspend fun deleteActivity(activity: Activity) {
        activityDao.delete(activity)
    }

    suspend fun loadMostRecentActivities(n: Int) = activityDao.loadMostRecentActivities(n)

    fun getActivity(id: Int) = activityDao.get(id)

    suspend fun getActivityImages() = activityDao.getImages()

    suspend fun startRecording(richRecording: RichRecording, context: Context) {
        val recordingId = recordingDao.insert(richRecording.recording).toInt()
        context.showRecordingNotification(
            recordingId,
            richRecording.type,
            richRecording.recording.startTime.time
        )
    }

    fun getRecordings() = recordingDao.getRecordings()

    suspend fun getRecording(uid: Int) = recordingDao.getById(uid)

    suspend fun deleteRecording(recording: Recording) {
        Log.d(TAG, "Deleting $recording")
        recording.uid?.let { context.hideRecordingNotification(it) }
        recordingDao.delete(recording)
    }

    suspend fun endRecording(recording: Recording) {
        Log.d(TAG, "Ending recording $recording")
        recording.uid?.let { context.hideRecordingNotification(it) }
        val type = getActivityType(recording.typeId)
        if (type == null) {
            Log.e(TAG, "Could not retrieve type for activity")
            return
        }
        val activity = recording.toActivity(type)
        activityDao.stopRecording(recording, activity)
    }

    suspend fun saveActivityType(activityType: ActivityType) = activityTypeDao.save(activityType)

    suspend fun deleteActivityType(activityType: ActivityType) =
        activityTypeDao.delete(activityType)

    suspend fun loadActivityTypes() = activityTypeDao.loadTypes()

    fun getActivityTypes() = activityTypeDao.getTypes()

    private suspend fun getActivityType(id: Int): ActivityType? = activityTypeDao.get(id)

    fun getActivityTypeRows() =
        getActivityTypesByCategory().map { ActivityTypeOrder.getRowsOrDefault(it) }

    private fun getActivityTypesByCategory() = activityTypeDao.getActivityTypesByCategory()

    suspend fun upsertFilterHistoryChips(chips: List<ActivityFilterChip>) =
        database.withTransaction {
            val historyItems = getFilterHistoryItems().first()
                .associate { it.toActivityFilterChip()!! to it.item.uid!! }
            for (item in chips) {
                val existingItemId = historyItems[item]
                filterHistoryDao.upsert(
                    item.toFilterHistoryItem().copy(uid = existingItemId)
                )
            }
            filterHistoryDao.onlyKeepNewest(8)
        }

    fun getFilterHistoryItems() = filterHistoryDao.getItems()

    fun getDays() = dayDao.getDays()

    fun getDay(day: EpochDay): Flow<Day> = dayDao.get(day).map { it ?: Day(day = day) }

    suspend fun saveDay(day: Day) {
        dayDao.upsert(day)
    }

    suspend fun savePlace(place: Place) = placeDao.upsert(place)

    suspend fun deletePlace(place: Place) = placeDao.delete(place)

    fun getPlaces() = placeDao.getAll()
}