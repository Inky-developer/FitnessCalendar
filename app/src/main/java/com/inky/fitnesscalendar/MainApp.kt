package com.inky.fitnesscalendar

import android.app.Application
import android.util.Log
import androidx.core.net.toUri
import com.inky.fitnesscalendar.di.ActivityTypeOrder
import com.inky.fitnesscalendar.di.DecisionTrees
import com.inky.fitnesscalendar.repository.AppRepository
import com.inky.fitnesscalendar.util.cleanImageStorage
import com.inky.fitnesscalendar.util.getOrCreateImagesDir
import com.inky.fitnesscalendar.util.getOrCreateSharedMediaCache
import dagger.hilt.android.HiltAndroidApp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltAndroidApp
class MainApp : Application() {
    @Inject
    lateinit var appRepository: AppRepository

    override fun onCreate() {
        super.onCreate()

        MainScope().launch(Dispatchers.IO) {
            initializeData()
            cleanupStorage()
        }
    }

    private suspend fun cleanupStorage() {
        val imageDir = getOrCreateImagesDir().toPath()
        val usedImages = appRepository.getUsedImages().map { imageDir.resolve(it).toFile().toUri() }
        cleanImageStorage(usedImages.toSet())

        getOrCreateSharedMediaCache().listFiles()?.forEach {
            it.delete()
        }
    }

    private suspend fun initializeData() {
        Log.i("MainApp", "Initializing app data")

        val activities = appRepository.loadMostRecentActivities(200)

        ActivityTypeOrder.init(activities)
        DecisionTrees.init(activities)
        Log.i("MainApp", "App data successfully initialized")
    }
}