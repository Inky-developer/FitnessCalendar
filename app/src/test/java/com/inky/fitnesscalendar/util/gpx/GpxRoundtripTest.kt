package com.inky.fitnesscalendar.util.gpx

import androidx.test.core.app.ApplicationProvider
import com.inky.fitnesscalendar.MainApp
import com.inky.fitnesscalendar.db.entities.Activity
import com.inky.fitnesscalendar.db.entities.RichActivity
import com.inky.fitnesscalendar.db.entities.Track
import com.inky.fitnesscalendar.testUtils.SportActivityType
import org.approvaltests.Approvals
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.io.ByteArrayInputStream
import java.io.StringWriter
import java.time.Instant
import java.util.Date

@RunWith(RobolectricTestRunner::class)
class GpxRoundtripTest {
    @Test
    fun testRoundtrip() {
        val context = ApplicationProvider.getApplicationContext<MainApp>()

        val tracks = GpxReader.read(ByteArrayInputStream(GPX.encodeToByteArray()))?.tracks
        requireNotNull(tracks)

        val gpxTrack = tracks[0]
        val track = Track(uid = 1, activityId = 1, points = gpxTrack.points)
        val activity = Activity(uid = 1, typeId = 1, startTime = Date.from(Instant.now()))
        val richActivity = RichActivity(activity, SportActivityType, null)

        val writer = StringWriter()
        GpxWriter.write(richActivity, track, context, writer)

        val newTracks =
            GpxReader.read(ByteArrayInputStream(writer.toString().encodeToByteArray()))?.tracks
        requireNotNull(newTracks)

        val newGpxTrack = newTracks[0]
        assertEquals(gpxTrack.points, newGpxTrack.points)

        Approvals.verifyAll(
            "gpx tracks names (old and new)",
            arrayOf(gpxTrack.name, newGpxTrack.name)
        )
    }
}