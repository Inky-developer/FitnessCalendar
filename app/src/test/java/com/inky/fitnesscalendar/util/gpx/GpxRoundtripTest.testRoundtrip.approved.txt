gpx tracks names (old and new)[0] = Cool track

Was a nice track
gpx tracks names (old and new)[1] = Sports
